sharding/handler/shard_tracker.py

```diff
@@ -1,107 +1,136 @@
+        else:
+            self.config = config
         self.shard_id = shard_id
         self.log_handler = log_handler
         self.smc_handler_address = smc_handler_address
-        self.current_score = None
-        self.new_logs = []
-        self.unchecked_logs = []
-
-    @to_tuple
-    def _get_new_logs(self):
-        shard_id_topic_hex = encode_hex(self.shard_id.to_bytes(32, byteorder='big'))
-        new_logs = self.log_handler.get_new_logs(
+
+    def _get_logs_by_shard_id(self, event_name, from_block=None, to_block=None):
+        """Search logs by the shard id.
+        """
+        return self.log_handler.get_logs(
             address=self.smc_handler_address,
             topics=[
-                encode_hex(COLLATION_ADDED_TOPIC),
-                shard_id_topic_hex,
+                encode_hex(get_event_signature_from_abi(event_name)),
+                encode_hex(self.shard_id.to_bytes(32, byteorder='big')),
             ],
+            from_block=from_block,
+            to_block=to_block,
+        )
+
+    def _get_logs_by_notary(self, event_name, notary, from_block=None, to_block=None):
+        """Search logs by notary address.
+
+        NOTE: The notary address provided must be padded to 32 bytes
+        and also hex-encoded. If notary address provided
+        is `None`, it will return all logs related to the event.
+        """
+        return self.log_handler.get_logs(
+            address=self.smc_handler_address,
+            topics=[
+                encode_hex(get_event_signature_from_abi(event_name)),
+                notary,
+            ],
+            from_block=from_block,
+            to_block=to_block,
+        )
+
+    #
+    # Basic functions to get emitted logs
+    #
+    @to_list
+    def get_register_notary_logs(self):
+        logs = self._get_logs_by_notary(event_name='RegisterNotary', notary=None)
+        for log in logs:
+            yield LogParser(event_name='RegisterNotary', log=log)
+
+    @to_list
+    def get_deregister_notary_logs(self):
+        logs = self._get_logs_by_notary(event_name='DeregisterNotary', notary=None)
+        for log in logs:
+            yield LogParser(event_name='DeregisterNotary', log=log)
+
+    @to_list
+    def get_release_notary_logs(self):
+        logs = self._get_logs_by_notary(event_name='ReleaseNotary', notary=None)
+        for log in logs:
+            yield LogParser(event_name='ReleaseNotary', log=log)
+
+    @to_list
+    def get_add_header_logs(self):
+        logs = self._get_logs_by_shard_id(event_name='AddHeader')
+        for log in logs:
+            yield LogParser(event_name='AddHeader', log=log)
+
+    @to_list
+    def get_submit_vote_logs(self):
+        logs = self._get_logs_by_shard_id(event_name='SubmitVote')
+        for log in logs:
+            yield LogParser(event_name='SubmitVote', log=log)
+
+    #
+    # Functions for user to check the status of registration or votes
+    #
+    def is_notary_registered(self, notary):
```

Comment:
It seems like this function only check for logs in the current period as they don't specify a `from_block` and give wrong results if called in later periods. Same with `is_notary_deregistered` and `is_notary_released`.

Reply:
Also: Is it possible that a notary registers and deregisters in the same period? If so, `is_notary_registered` should probably check for `DeregisterNotary` events as well.

Reply:
> It seems like this function only check for logs in the current period as they don't specify a from_block and give wrong results if called in later periods.

Yes you are right. They should be able to specify `from_block` if they want to query this over a certain span of periods.